doc/source/fault-tolerance.rst

```diff
@@ -1,50 +1,91 @@
+            print('SUCCESS')
+        except ray.exceptions.RayWorkerError:
+            print('FAILURE')
 
-If an object is needed but is lost or was never created, then the task that
-created the object will be re-executed to create the object. If necessary, tasks
-needed to create the input arguments to the task being re-executed will also be
-re-executed. This is the standard *lineage-based fault tolerance* strategy used
-by other systems like Spark.
 
 Actors
 ------
 
-When an actor dies (either because the actor process crashed or because the node
-that the actor was on died), by default any attempt to get an object from that
-actor that cannot be created will raise an exception. Subsequent releases will
-include an option for automatically restarting actors.
+If an actor process crashes unexpectedly, Ray will attempt to reconstruct the
+actor process up to a maximum number of times. This value can be specified with
+the ``max_reconstructions`` keyword, which by default is ``0``. If the maximum
+number of reconstructions has been used up, then subsequent actor methods will
+raise exceptions.
+
+When an actor is reconstructed, its state will be recreated by rerunning its
+constructor.
+
+You can experiment with this behavior by running the following code.
+
+.. code-block:: python
 
-Current Limitations
--------------------
+    import numpy as np
+    import os
+    import ray
+    import time
 
-At the moment, Ray does not handle all failure scenarios. We are working on
-addressing these known problems.
+    ray.init(ignore_reinit_error=True)
 
-Process Failures
-~~~~~~~~~~~~~~~~
+    @ray.remote(max_reconstructions=5)
+    class Actor:
+        def __init__(self):
+            self.counter = 0
 
-1. Ray does not recover from the failure of any of the following processes:
-   any of the Redis servers and the monitor process.
-2. If a driver fails, that driver will not be restarted and the job will not
-   complete.
+        def increment_and_possibly_fail(self, failure_probability):
+            self.counter += 1
+            time.sleep(0.2)
+            if np.random.random() < failure_probability:
+                os._exit(0)
+            return self.counter
 
-Lost Objects
-~~~~~~~~~~~~
+    actor = Actor.remote()
 
-1. If an object is constructed by a call to ``ray.put`` on the driver, is then
-   evicted, and is later needed, Ray will not reconstruct this object.
-2. If an object is constructed by an actor method, is then evicted, and is later
-   needed, Ray will not reconstruct this object.
+    # The actor will be reconstructed up to 5 times. After that, methods will
+    # raise exceptions. The actor is reconstructed by rerunning its
+    # constructor. Methods that were executing when the actor died will also
+    # raise exceptions.
+    for _ in range(100):
+        try:
+            counter = ray.get(actor.increment_and_possibly_fail.remote(0.1))
+            print(counter)
+        except ray.exceptions.RayActorError:
```

Comment:
I would expect no exceptions to be raised until `max_reconstructions` has been exceeded. Instead, it looks like methods that are running when the actor dies also raise exceptions. Is that intentional? @edoakes @stephanie-wang @raulchen

Reply:
This seems easy to change, but we would need some way of detecting reconstruct if it's not exposed as an error

Reply:
It shouldn't raise exceptions in that case. Reconstruction should be transparent to the caller. Can you make this example deterministic (as I mentioned in the above thread) to confirm this behavior?

Reply:
@raulchen I just made it deterministic. If you try running it, you'll see that it raises exceptions way before the max_reconstructions have been used up.

Reply:
Right now we only do resubmission for non-actor tasks, thus if the remote actor dies, tasks that haven't finished execution would be marked as failed. Or do we want to support re-submission for actor tasks as well?

Reply:
oh, I see. Re-submission isn't supported for direct actors yet. For old non-direct actors, reconstruction is completely transparent.

Reply:
Once that changes, we should update this documentation.